### PR TITLE
Support multiple targets for checkOnSave (in conjunction with cargo 1.64.0+)

### PR DESCRIPTION
**Pull request by poliorcetics**
This PR adds support for the ability to pass multiple `--target` flags when using
`cargo` 1.64.0+.

## Questions

I needed to change the type of two configurations options, but I did not plurialize the names to
avoid too much churn, should I ?

## Zulip thread

https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Issue.2013282.20.28supporting.20multiple.20targets.20with.201.2E64.2B.29

## Example

To see it working, on a macOS machine:

```sh
$ cd /tmp
$ cargo new cargo-multiple-targets-support-ra-test
$ cd !$
$ mkdir .cargo
$ echo '
[build]
target = [
    "aarch64-apple-darwin",
    "x86_64-apple-darwin",
]
' > .cargo/config.toml
$ echo '
fn main() {
    #[cfg(all(target_arch = "aarch64", target_os = "macos"))]
    {
        let a = std::fs::read_to_string("/tmp/test-read");
    }

    #[cfg(all(target_arch = "x86_64", target_os = "macos"))]
    {
        let a = std::fs::read_to_string("/tmp/test-read");
    }

    #[cfg(all(target_arch = "x86_64", target_os = "windows"))]
    {
        let a = std::fs::read_to_string("/tmp/test-read");
    }
}
' > src/main.rs
# launch your favorite editor with the version of RA from this PR
#
# You should see warnings under the first two `let a = ...` but not the third
```

## Screen

Two panes of a terminal emulator, on the left pane is the main.rs file described above, with warnings for the first two let a = declaration, on the right pane is a display of the .cargo/config.toml, an ls of the current files in the directory and a call to cargo build to show the same warnings as in the editor on the left pane

Helps with #13282

### CODE REVIEW

File: crates/project-model/src/cargo_workspace.rs [hunk 1]
```diff
@@ -466,10 +466,23 @@ impl CargoWorkspace {
...
+    }
+
+    rustc_discover_host_triple(cargo_toml, &config.extra_env)
+}
+
 fn rustc_discover_host_triple(
     cargo_toml: &ManifestPath,
     extra_env: &FxHashMap<String, String>,
-) -> Option<String> {
+) -> Vec<String> {
```
Comment (bjorn3) [inline crates/project-model/src/cargo_workspace.rs:485]: I think this should stay `Option<String>`. There can only be 1 host triple.

Comment (poliorcetics) [reply crates/project-model/src/cargo_workspace.rs]: Done

**Comment by poliorcetics** [comment]
I didn't make the change for the `rust-project.json` config because `rustc` does not accept multiple `--target` flags, contrarily to `cargo`

**Comment by flodiebold** [comment]
I don't think we can support this for `cargo.target`, at most for checkOnSave.

**Comment by poliorcetics** [comment]
> I don't think we can support this for `cargo.target`, at most for checkOnSave.

`cargo` on 1.64.0+ accepts multiple targets, it's even supported in the `.cargo/config.toml`, why would `cargo.target` not support it ?

**Comment by flodiebold** [comment]
Because `cargo.target` controls our analysis, and that doesn't support multiple targets (without a lot more code).

**Comment by poliorcetics** [comment]
> Because `cargo.target` controls our analysis, and that doesn't support multiple targets (without a lot more code).

Oh, I'll try to do it, either in this PR if it's not merged by then, or in another. For now, I removed it from the commit and changed back to an `Option`.

**Comment by poliorcetics** [comment]
## Progress report

I made some progress on this today 🎉.

I have completions working for several targets at once:

std::arch::x86_64::__m12 correctly proposes the completion for __m128

**But** (because of course it's not perfectly working), completions for conflicting names fail, I don't which is chosen first (I suspect the first configured target): 

sys_func(usize, usize) is selected instead of the single-argument variant, which is the one to use on x86_64-apple-darwin

**Comment by poliorcetics** [comment]
`Go to definition` and errors will point to the wrong item with the implementation I used, I removed that part and kept only the multiple targets when running `cargo`, not inside R-A

File: crates/rust-analyzer/src/config.rs [hunk 2]
```diff
@@ -171,9 +173,9 @@ config_data! {
         /// ```
         /// .
         checkOnSave_overrideCommand: Option<Vec<String>> = "null",
-        /// Check for a specific target. Defaults to
+        /// Check for specific targets. Defaults to
         /// `#rust-analyzer.cargo.target#`.
-        checkOnSave_target: Option<String>               = "null",
+        checkOnSave_target: Vec<String>                  = "[]",
```
Comment (Veykril) [inline crates/rust-analyzer/src/config.rs:178]: We should still accept non arrays as well, easiest would probably be to use a `Vec<String>` wrapper struct here and use the `single_or_array` function for deserialization thats already defined in the module

Comment (poliorcetics) [reply crates/rust-analyzer/src/config.rs]: Fixed

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Pull request merged**